### ISSUE TITLE
No products generated, start/stop outside data takes

### ISSUE THREAD

**Issue by lmar76** [stcorp/procsim#14]
I'm trying to simulate L0 step 2 processing.

In this kind of processing, the CPF sets the start/stop times in the job order using the theoretic start/stop times of the slice to be processed, even if the slice is partial (i.e. the data take starts after slice theoretic start or stops before slice theoretic stop).

In the scenario being simulated, slice 2 have the following start/stop times:

[2017-03-02T06:07:21.530,2017-03-02T06:09:08.546]

but the data take starts at 2017-03-02T**06:07:23.674** (i.e. partial slice).

I have executed test with the working directory and scenario configuration included in [L0PFS2_slice2_WD_scenario.zip](https://github.com/stcorp/procsim/files/7538527/L0PFS2_slice2_WD_scenario.zip).

The output log is the following:

```
2021-11-15T12:54:20.943310 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [I] Procsim v1.4.0 processor stub simulator
2021-11-15T12:54:20.943377 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [I] Simulate scenario LOAD_TDS - SM idx 0 DTID 7217244 L0PFS2
2021-11-15T12:54:20.943389 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [I] Read JobOrder BIO_L0PFS2_01.00_7217244-02_20211112T120000_JobOrder.xml
2021-11-15T12:54:20.943397 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [D] JobOrder validation against schema: OK
2021-11-15T12:54:20.943405 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [I] Read task L0PFS2 from the JobOrder
2021-11-15T12:54:20.943414 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [I] Processing parameter dataTakeID = 7217244
2021-11-15T12:54:20.943425 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [I] Input type RAWS026_10: BIO_RAWS025_10_20170302T060723_20170302T060733_D20170302T073617_01_BEU0LE
2021-11-15T12:54:20.943433 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [I] Input type RAWS026_10: BIO_RAWS025_10_20170302T060723_20170302T060908_D20170302T073617_01_BEU0LE
2021-11-15T12:54:20.943441 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [I] Input type RAWS026_10: BIO_RAWS025_10_20170302T060856_20170302T060937_D20170302T073617_01_BEU0LE
2021-11-15T12:54:20.943448 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [I] Input type RAWS026_10: BIO_RAWS026_10_20170302T060723_20170302T060733_D20170302T073617_01_BEU0M9
2021-11-15T12:54:20.943631 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [I] Input type RAWS026_10: BIO_RAWS026_10_20170302T060723_20170302T060908_D20170302T073617_01_BEU0M9
2021-11-15T12:54:20.943657 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [I] Input type RAWS026_10: BIO_RAWS026_10_20170302T060856_20170302T060937_D20170302T073617_01_BEU0M9
2021-11-15T12:54:20.943673 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [I] Input type RAWS023_10: BIO_RAWS023_10_20170302T060723_20170302T060733_D20170302T073617_01_BEU0JG
2021-11-15T12:54:20.943685 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [I] Input type RAWS023_10: BIO_RAWS023_10_20170302T060723_20170302T060908_D20170302T073617_01_BEU0JG
2021-11-15T12:54:20.943693 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [I] Input type RAWS023_10: BIO_RAWS023_10_20170302T060856_20170302T060937_D20170302T073617_01_BEU0JG
2021-11-15T12:54:20.943701 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [I] Input type RAWS024_10: BIO_RAWS024_10_20170302T060723_20170302T060733_D20170302T073617_01_BEU0KG
2021-11-15T12:54:20.943708 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [I] Input type RAWS024_10: BIO_RAWS024_10_20170302T060723_20170302T060733_D20170302T073617_01_BEU0KG
2021-11-15T12:54:20.943715 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [I] Input type RAWS024_10: BIO_RAWS024_10_20170302T060856_20170302T060937_D20170302T073617_01_BEU0KG
2021-11-15T12:54:20.943723 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [I] Input type AUX_PP0___: BIO_AUX_PP0____20000101T000000_20500101T000000_01_000000
2021-11-15T12:54:20.943731 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [I] Input type AUX_ERP___: BIO_AUX_ERP____20170302T000000_20170501T000000_01_91M800
2021-11-15T12:54:20.943739 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [I] Input type MPL_ORBPRE: BIO_TEST_MPL_ORBPRE_20170101T060001_20170910T063749_0001.EOF
2021-11-15T12:54:20.951137 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [I] Cannot "merge" incomplete H or V slices, no data from input files available. Set metadata_source in output section.
2021-11-15T12:54:20.951170 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [D] Set num_l0_lines to 314841
2021-11-15T12:54:20.951182 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [D] Set num_l0_lines_corrupt to 0
2021-11-15T12:54:20.951191 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [D] Set num_l0_lines_missing to 0
2021-11-15T12:54:20.951199 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [D] Set mission_phase to Interferometric
2021-11-15T12:54:20.951209 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [D] Overwrite global_coverage_id from NA to 01
2021-11-15T12:54:20.951217 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [D] Overwrite major_cycle_id from 1 to 02
2021-11-15T12:54:20.951224 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [D] Overwrite repeat_cycle_id from __ to 03
2021-11-15T12:54:20.951231 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [D] Overwrite track_nr from 0 to 1
2021-11-15T12:54:20.951239 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [D] Set slice_frame_nr to 2
2021-11-15T12:54:20.951249 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [D] Overwrite begin_end_position_from_toi from False to True
2021-11-15T12:54:20.951598 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [D] Output directory is /opt/pnode/data/WD-L0PFS2-7217244-02/outputs
2021-11-15T12:54:20.951620 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [D] Use begin_position as input for validity start time
2021-11-15T12:54:20.951629 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [D] Use end_position as input for validity stop time
2021-11-15T12:54:20.951694 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [I] No products generated, start/stop outside data takes?
2021-11-15T12:54:20.951955 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [I] Cannot "merge" incomplete H or V slices, no data from input files available. Set metadata_source in output section.
2021-11-15T12:54:20.951979 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [D] Set num_l0_lines to 314841
2021-11-15T12:54:20.951990 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [D] Set num_l0_lines_corrupt to 0
2021-11-15T12:54:20.951998 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [D] Set num_l0_lines_missing to 0
2021-11-15T12:54:20.952007 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [D] Set mission_phase to Interferometric
2021-11-15T12:54:20.952016 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [D] Overwrite global_coverage_id from NA to 01
2021-11-15T12:54:20.952028 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [D] Overwrite major_cycle_id from 1 to 02
2021-11-15T12:54:20.952042 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [D] Overwrite repeat_cycle_id from __ to 03
2021-11-15T12:54:20.952057 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [D] Overwrite track_nr from 0 to 1
2021-11-15T12:54:20.952072 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [D] Set slice_frame_nr to 2
2021-11-15T12:54:20.952090 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [D] Overwrite begin_end_position_from_toi from False to True
2021-11-15T12:54:20.952175 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [D] Output directory is /opt/pnode/data/WD-L0PFS2-7217244-02/outputs
2021-11-15T12:54:20.952223 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [D] Use begin_position as input for validity start time
2021-11-15T12:54:20.952234 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [D] Use end_position as input for validity stop time
2021-11-15T12:54:20.952299 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [I] No products generated, start/stop outside data takes?
2021-11-15T12:54:20.952317 node-name L0PFS2 01.00 L0PFS2 [0000007073]: [I] Task done, exit with code 0
```

As you can see, Procism logs `[I] No products generated, start/stop outside data takes?` and no products are generated.

To allow Procsim to generate products I have to update the start/stop time in the Job Order as follows:

```xml
<Request>
  <TOI>
    <Start>2017-03-02T06:07:23.674000</Start>
    <Stop>2017-03-02T06:09:08.545000</Stop>
  </TOI>
</Request>
```

I propose to change this behavior such that Procsim emits the warning:  `[W] start/stop outside data takes?`  (please note the changed logging level) and generates the output using the intersection (i.e. the overlap interval) between the TOI in the Job Order and the data take interval.

Do you agree?

**Comment by mivermeu** [comment]
Indeed, this was an oversight that has been fixed with PR #18. In the even that the data take start occurs in the middle of a slice, the product generation start time is simply shifted to the data take start time.

**Comment by lmar76** [comment]
@mivermeu thank you, it works!
